### ISSUE TITLE
[Performance] Replace Naive Quantized Convolution with Optimized GEMM-based Implementation

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#345]
# [Performance] Replace Naive Quantized Convolution with Optimized GEMM-based Implementation

## Problem Description

The `conv2d_quantized` function in `crates/bitnet-kernels/src/convolution.rs` currently uses a naive nested-loop implementation that performs quantized convolution with on-the-fly dequantization. This approach is inefficient for production workloads and represents a simulation rather than an optimized kernel implementation.

## Environment

- **File**: `/home/steven/code/Rust/BitNet-rs/crates/bitnet-kernels/src/convolution.rs`
- **Function**: `conv2d_quantized` (lines 194-362)
- **Codebase**: BitNet.rs neural network inference framework
- **Impact**: All quantized convolution operations in BitNet models
- **Feature Gates**: Affects both `cpu` and `gpu` feature builds

## Current Implementation Analysis

The current implementation uses 8 nested loops with on-the-fly dequantization:

```rust
// Perform quantized convolution with dequantization on-the-fly
for batch in 0..n {
    for out_ch in 0..oc {
        let scale = weight_scales[out_ch];
        for in_ch in 0..ic {
            for y in 0..oh {
                for x in 0..ow {
                    for ky in 0..kh {
                        for kx in 0..kw {
                            // Dequantize weight for each access
                            let weight_val = dequantize_weight(
                                weight_quantized,
                                weight_linear_idx,
                                qtype,
                                scale,
                            )?;
                            output[output_idx] += input[input_idx] * weight_val;
                        }
                    }
                }
            }
        }
    }
}
```

**Performance Issues:**
1. **O(N×C×H×W×Kh×Kw) dequantization calls**: Each weight is dequantized multiple times
2. **Poor cache locality**: Nested loops create suboptimal memory access patterns
3. **No SIMD utilization**: Sequential processing prevents vectorization
4. **Branch overhead**: Quantization type checking in inner loops

## Root Cause Analysis

The implementation prioritizes simplicity over performance, treating quantized convolution as a direct analog of FP32 convolution with value substitution. This misses opportunities for:

1. **Batch dequantization**: Dequantize weight blocks once and reuse
2. **GEMM transformation**: Convert convolution to optimized matrix multiplication
3. **Kernel integration**: Leverage existing `KernelProvider` infrastructure for device-specific optimizations
4. **Memory optimization**: Reduce temporary allocations and improve access patterns

## Impact Assessment

**Severity**: High - Core performance bottleneck
**Affected Components**:
- All BitNet models using quantized convolution layers
- Inference latency for CNN-based architectures
- GPU utilization efficiency (when feature enabled)
- Cross-validation accuracy vs. reference implementations

**Performance Impact**:
- Estimated 5-20x slower than optimized implementation
- Scales poorly with model size and batch dimensions
- Blocks production deployment for convolutional models

## Proposed Solution

Implement optimized quantized convolution using GEMM-based approach with device-aware kernel selection:

### Primary Implementation Strategy

1. **Quantized Im2Col Transformation**
   ```rust
   fn quantized_im2col(
       input: &[f32],
       weight_quantized: &[u8],
       weight_scales: &[f32],
       qtype: QuantizationType,
       // ... dimensions and params
   ) -> Result<QuantizedGemmData>
   ```

2. **Batch Weight Dequantization**
   ```rust
   fn batch_dequantize_weights(
       weight_quantized: &[u8],
       weight_scales: &[f32],
       qtype: QuantizationType,
       output_channel: usize,
   ) -> Result<Vec<f32>>
   ```

3. **Device-Aware Kernel Integration**
   ```rust
   pub fn conv2d_quantized(/* ... */) -> Result<()> {
       let kernel_manager = KernelManager::new();
       let provider = kernel_manager.select_best()?;

       // Transform to GEMM format
       let gemm_data = quantized_im2col(input, weight_quantized, /* ... */)?;

       // Dispatch to optimized matmul
       provider.matmul_quantized(&gemm_data.input, &gemm_data.weights, output, /* ... */)
   }
   ```

### Alternative Approaches (Future Considerations)

1. **Direct Quantized GEMM**: Extend `KernelProvider::matmul_i2s` for convolution workloads
2. **Winograd Quantized**: Specialized algorithm for small kernels (3x3, 5x5)
3. **FFT Quantized**: For large kernel convolutions (rare in BitNet models)

### Phase 1: Core Infrastructure (1-2 weeks)
- [ ] Add `matmul_quantized` method to `KernelProvider` trait
- [ ] Implement quantized im2col transformation utility
- [ ] Create batch weight dequantization functions
- [ ] Add comprehensive dimension validation

### Phase 2: Kernel Integration (1 week)
- [ ] Extend `FallbackKernel` with quantized GEMM support
- [ ] Update device-aware selection logic for quantized operations
- [ ] Implement memory-efficient workspace allocation

### Phase 3: Optimization & Validation (1 week)
- [ ] Add CPU SIMD optimizations (AVX2/AVX-512/NEON paths)
- [ ] Implement GPU CUDA kernel integration
- [ ] Cross-validate against existing naive implementation
- [ ] Performance benchmarking and regression testing

### Phase 4: Production Integration (1 week)
- [ ] Update quantization type support (I2_S, TL1, TL2)
- [ ] Add comprehensive error handling and edge cases
- [ ] Documentation and usage examples
- [ ] Integration testing with real BitNet models

### Correctness Validation
- [ ] **Numerical equivalence**: Output matches current implementation within tolerance
- [ ] **Cross-validation**: Results consistent with C++ reference when available
- [ ] **Edge case testing**: Boundary conditions, unusual dimensions, padding scenarios
- [ ] **Quantization accuracy**: Verify dequantization correctness across all QuantizationType variants

### Performance Testing
- [ ] **Latency benchmarks**: Measure speedup vs. current implementation
- [ ] **Memory profiling**: Validate reduced allocation overhead
- [ ] **Device scaling**: Test performance across CPU/GPU architectures
- [ ] **Model integration**: End-to-end inference timing with real BitNet models

### Regression Testing
- [ ] **Feature gate compatibility**: Ensure correct builds with `--features cpu` and `--features gpu`
- [ ] **Cross-compilation**: Validate ARM64, x86_64 targets
- [ ] **Existing test suite**: All current convolution tests continue passing

### Functional Requirements
- ✅ **API Compatibility**: Existing `conv2d_quantized` function signature unchanged
- ✅ **Numerical Accuracy**: Output matches current implementation within 1e-6 tolerance
- ✅ **Feature Support**: All quantization types (I2_S, TL1, TL2) correctly handled
- ✅ **Error Handling**: Maintains existing error conditions and validation

### Performance Requirements
- ✅ **Latency Improvement**: Minimum 3x speedup for typical convolution sizes
- ✅ **Memory Efficiency**: Reduced peak memory usage vs. current implementation
- ✅ **Device Optimization**: Leverages available SIMD/GPU acceleration
- ✅ **Scalability**: Performance improvement increases with problem size

### Quality Requirements
- ✅ **Cross-validation**: Passes all existing convolution tests
- ✅ **Documentation**: Implementation approach documented in code comments
- ✅ **Integration**: Works seamlessly with existing `KernelManager` infrastructure
- ✅ **Maintainability**: Code structure follows BitNet.rs patterns and conventions

## Related Issues

- **Related**: Naive `conv2d` implementation optimization (separate issue for FP32 version)
- **Depends on**: `KernelProvider` trait extension for quantized operations
- **Blocks**: Production deployment of quantized convolutional models
- **Cross-ref**: GPU kernel architecture improvements for convolution workloads

### Quantization Type Handling
- **I2_S**: 2-bit signed with 4 values per byte packing
- **TL1/TL2**: 8-bit table lookup with different dequantization strategies
- **Future**: IQ2_S GGML compatibility via FFI bridge

### Memory Layout Considerations
- **Input**: NCHW format (batch, channels, height, width)
- **Weights**: OIHW format (out_channels, in_channels, height, width)
- **Im2Col Output**: Efficient column-major layout for GEMM operations
- **Workspace**: Temporary allocation strategy for intermediate results

### Device-Specific Optimizations
- **CPU**: Leverage existing AVX2/AVX-512/NEON matmul kernels
- **GPU**: CUDA quantized GEMM with Tensor Core utilization when available
- **Fallback**: Maintain current naive implementation as reference/compatibility layer

---

**Labels**: `performance`, `kernels`, `quantization`, `convolution`, `high-priority`
**Assignee**: TBD
**Milestone**: Production Performance Optimization
**Priority**: High - Core inference performance bottleneck
